openapi.yml

```diff
@@ -12,6 +12,14 @@ paths:
       responses:
         201:
```

Comment:
Yeah, if this would help the front-end, we could return the orderdigest as UUID. Sure

Reply:
If we get a UUID per order|orders, it would be super helpful to also have an endpoint to check on the status of an order.
Like `api/v1/order/<uuid> or even api/v1/order?uuid=[...ids]`
Or do you feel `/api/v1/orders?owner=..&sellToken=..&buyToken=..` already serves that purpose

Easily checking for individual order status by uuid would be nicer to use when a user has many orders historically but only a few pending.

Reply:
I don't think orders have a specific uuid. It's just the hash and you already know this in the frontend because the user had to sign the order before submitting.

For the status it's not something i've thought about before. It depends on how the contract works and what the orderbook service is responsible for. Initially I thought it was really only there to collect and distribute orders and wouldn't even need to talk to an ethereum node, in which case there is no status: an order either exists or doesn't. I'm guessing an additional status would be whether it got matched but I'm not sure whether this is something that the orderbook service should be responsible for.

Reply:
I think it's a good idea, eventually, to allow the API be as close as possible as other exchanges, and be the only requirement for a CLI/UI to trade in GPv2

Usually all exchanges allow this, and I think would make integrations way easier. I believe this could help a lot adoption, and integrations with tools. 

So, regarding the state, sure they can review all settlements and try to identify the orders there and guess the state. It will get complicated when the settlement failed to execute the order specially if here's re-submissions.
So I find incredibly useful for a client to get an ID back and being able to ask for just the state of it. in summary I would focus in this product in making it look SIMPLER first. 
In this direction, one initial proposal for the state transitions (focusing on the most important use cases):


I understand a random UUID cannot be used, because this would limit in the future having a pool of operators you could use. It needs to be deterministic, but I would make the clients agnostic to how you generate the identity of the order. A hash of it's content sounds good to me.

Reply:
What you describe sounds useful but should be a separate service because it does many more things than this orderbook part of the system. I would like this to really only be the exact functionality it was intended for which is receiving raw orders and distributing them to solvers.

Reply:
I think the service could be different, if you believe is fundamentally different responsibilities. I can also see that. Maybe even an implementation of this service, could just be either a custom indexer or the graph. 

However, I think that regardless of the logical separations in the modules, we should provide a consistent simple API, just for making it simpler for others.

Reply:
The order service will be a single point of failure - at least it will not be decentralized in the beginning - and hence, I would prefer to keep it as simple as possible. 
Yes, spinning up another service for the order status, will definitively be needed. Let's discuss this server later in another PR.

Yes, the UUID would be the hash=orderdigest. I agree with Valentin that the front-end should be already aware of it.

Reply:
> The order service will be a single point of failure - at least it will not be decentralized in the beginning - and hence, I would prefer to keep it as simple as possible.

Sure, I think I'm happy with the separation. I think it's fine to be independent and under the same URL

> Yes, the UUID would be the hash=orderdigest. I agree with Valentin that the front-end should be already aware of it.

What do you mean with this? You referring to aware as generating himself, or aware that he receives the ID from the creation endpoint. I think for the base use case, it's better to just return it so u don't need to explain anything else to someone integrating. On top of that, yes, it's nice that our JS library also has a function that returns the ID for an order, but a basic integration wouldn't use that imo.

Reply:
> What do you mean by this?

https://github.com/gnosis/oba-contracts/blob/7adfd42f6a5e0735b318cb2dd3928198656bd384/src/ts/index.ts#L311-L322

Orders have a well-defined hash (EIP-712) that uniquely identifies them. I think what @e00E and @josojo are saying is that this needs to be computed in the FE anyway as this hash is used for signing the order.

One small detail, however, is that this hashing is typically done by the signing wallet with EIP-712. However, I do think `web3` libraries typically implement EIP-712 struct hashes.

> the UUID would be the hash=orderdigest

One issue is that this is not unique for all orders, but rather unique for all orders of a given user. In that sense an order UUID is: `(ownerAddress, orderDigest)`. Alternatively we can use the `orderKey` as the UUID which is `keccak256(orderDitest, ownerAddress, validTo)`.

Reply:
Yes, right. Using the `orderKey` as UUID would be better than orderDigest. If the endpoint should return it, we can do it. If FE wants to compute it itself, this is also fine for me. Both ways are reasonable IMO.

Reply:
Cool, thanks for clarifying. Yep, I think then we should aim for both :) 

> One small detail, however, is that this hashing is typically done by the signing wallet with EIP-712. However, I do think web3 libraries typically implement EIP-712 struct hashes.

Also, there would be potentially sth weird with if you send the same order, with no expiration date in two different moments in time

I think, it would be good to include a `validFrom`, right? (not just for this, but also)

Reply:
We don't have a `validFrom` in the spec anymore. To our understanding, this feature was not used much besides the IDO usecase and it would consume further gas.

>Yep, I think then we should aim for both :)

Okay, then I will change the API to return the orderKey